In [1]:
import pandas as pd

pd.set_option('display.max_columns', 30)

# Invoice Generation from CUR

In [2]:
catalog.list()

['accounts_per_organization',
 's3_pocs_cur_dataset',
 's3_intelbras_cur_dataset',
 's3_infra2_cur_dataset',
 'pocs_cur_dataset',
 'intelbras_cur_dataset',
 'infra2_cur_dataset',
 'pocs_cur_dataset_enriched',
 'intelbras_cur_dataset_enriched',
 'infra2_cur_dataset_enriched',
 'parameters',
 'params:data_prep_raw_params',
 'params:data_prep_raw_params.kept_columns']

## Analyzing Infra 2 CUR of 2023-05

## Loading from S3

In [3]:
s3_infra2_cur = list(catalog.load('s3_infra2_cur_dataset').items())
s3_infra2_cur202305 = [t for t in s3_infra2_cur if t[0] == 'year=2023/month=5/CUR-00001.snappy.parquet'][0]
infra2_cur202305 = s3_infra2_cur202305[1]()
infra2_cur202305

[07/06/23 18:44:14] INFO     Loading data from 's3_infra2_cur_dataset' (PartitionedDataSet)...  ]8;id=289874;file:///home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-packages/kedro/io/data_catalog.py\data_catalog.py]8;;\:]8;id=346057;file:///home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-packages/kedro/io/data_catalog.py#345\345]8;;\

identity_line_item_id  \
0     xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...   
1     xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...   
2     xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...   
3     xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...   
4     xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...   
...                                                 ...   
5509  leyy3urvpooyfs3iai36yaekmq4vxyvzlqzyyk47um2mbo...   
5510  ulwrbyeylpqfhj4zzwesf6z7ypcbcjenrqga4kidny3363...   
5511  ulwrbyeylpqfhj4zzwesf6z7ypcbcjenrqga4kidny3363...   
5512  ulwrbyeylpqfhj4zzwesf6z7ypcbcjenrqga4kidny3363...   
5513  ulwrbyeylpqfhj4zzwesf6z7ypcbcjenrqga4kidny3363...   

                         identity_time_interval bill_invoice_id  \
0     2023-05-01T00:00:00Z/2023-05-01T01:00:00Z      1349042489   
1     2023-05-01T01:00:00Z/2023-05-01T02:00:00Z      1349042489   
2     2023-05-01T02:00:00Z/2023-05-01T03:00:00Z      1349042489   
3     2023-05-01T03:00:00Z/2023-05-01T04:00:00Z      1349042489   
4     2023-05-01T04:00:00Z/2023-05-01T05:00:00Z      1349042489   
...                                         ...             ...   
5509  2023-05-31T14:00:00Z/2023-05-31T15:00:00Z      1349042489   
5510  2023-05-31T15:00:00Z/2023-05-31T16:00:00Z      1349042489   
5511  2023-05-31T16:00:00Z/2023-05-31T17:00:00Z      1349042489   
5512  2023-05-31T17:00:00Z/2023-05-31T18:00:00Z      1349042489   
5513  2023-05-31T18:00:00Z/2023-05-31T19:00:00Z      1349042489   

          bill_invoicing_entity bill_billing_entity bill_bill_type  \
0     Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
1     Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
2     Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
3     Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
4     Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
...                         ...                 ...            ...   
5509  Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
5510  Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
5511  Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
5512  Amazon Web Services, Inc.     AWS Marketplace    Anniversary   
5513  Amazon Web Services, Inc.     AWS Marketplace    Anniversary   

     bill_payer_account_id bill_billing_period_start_date  \
0             981064078967                     2023-05-01   
1             981064078967                     2023-05-01   
2             981064078967                     2023-05-01   
3             981064078967                     2023-05-01   
4             981064078967                     2023-05-01   
...                    ...                            ...   
5509          981064078967                     2023-05-01   
5510          981064078967                     2023-05-01   
5511          981064078967                     2023-05-01   
5512          981064078967                     2023-05-01   
5513          981064078967                     2023-05-01   

     bill_billing_period_end_date line_item_usage_account_id  \
0                      2023-06-01               273762107511   
1                      2023-06-01               273762107511   
2                      2023-06-01               273762107511   
3                      2023-06-01               273762107511   
4                      2023-06-01               273762107511   
...                           ...                        ...   
5509                   2023-06-01               167002523613   
5510                   2023-06-01               167002523613   
5511                   2023-06-01               167002523613   
5512                   2023-06-01               167002523613   
5513                   2023-06-01               167002523613   

     line_item_line_item_type line_item_usage_start_date  \
0                       Usage        2023-05-01 00:00:00   
1                       Usage        2023-05-01 01:00:00 

## Loading CUR1 from 01_raw

In [15]:
infra2_cur = list(catalog.load('infra2_cur_dataset').items())
infra2_cur

[07/06/23 18:49:34] INFO     Loading data from 'infra2_cur_dataset' (PartitionedDataSet)...     ]8;id=246454;file:///home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-packages/kedro/io/data_catalog.py\data_catalog.py]8;;\:]8;id=639530;file:///home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-packages/kedro/io/data_catalog.py#345\345]8;;\

[('2023_05_CUR-00001.csv',
  <bound method AbstractVersionedDataSet.load of <kedro_datasets.pandas.csv_dataset.CSVDataSet object at 0x7f5bd6206ad0>>),
 ('2023_05_CUR-00002.csv',
  <bound method AbstractVersionedDataSet.load of <kedro_datasets.pandas.csv_dataset.CSVDataSet object at 0x7f5bdc03fe80>>),
 ('2023_05_CUR-00003.csv',
  <bound method AbstractVersionedDataSet.load of <kedro_datasets.pandas.csv_dataset.CSVDataSet object at 0x7f5bdc03f460>>),
 ('2023_05_CUR-00004.csv',
  <bound method AbstractVersionedDataSet.load of <kedro_datasets.pandas.csv_dataset.CSVDataSet object at 0x7f5bd598d7b0>>),
 ('2023_05_CUR-00005.csv',
  <bound method AbstractVersionedDataSet.load of <kedro_datasets.pandas.csv_dataset.CSVDataSet object at 0x7f5bd59a60b0>>),
 ('2023_05_CUR-00006.csv',
  <bound method AbstractVersionedDataSet.load of <kedro_datasets.pandas.csv_dataset.CSVDataSet object at 0x7f5bd59a5de0>>),
 ('2023_05_CUR-00007.csv',
  <bound method AbstractVersionedDataSet.load of <kedro_datasets.pa

In [16]:
infra2_cur1 = infra2_cur[0][1]()
infra2_cur1.head(100)

identity_line_item_id  bill_invoice_id  \
0   xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
1   xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
2   xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
3   xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
4   xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
..                                                ...              ...   
95  xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
96  xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
97  xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
98  xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   
99  xcjvdeganjbzg4vdht3lnj77dmvvvs4hmipmmjrpowbmsl...       1349042489   

   bill_billing_entity      bill_invoicing_entity  \
0      AWS Marketplace  Amazon Web Services, Inc.   
1      AWS Marketplace  Amazon Web Services, Inc.   
2      AWS Marketplace  Amazon Web Services, Inc.   
3      AWS Marketplace  Amazon Web Services, Inc.   
4      AWS Marketplace  Amazon Web Services, Inc.   
..                 ...                        ...   
95     AWS Marketplace  Amazon Web Services, Inc.   
96     AWS Marketplace  Amazon Web Services, Inc.   
97     AWS Marketplace  Amazon Web Services, Inc.   
98     AWS Marketplace  Amazon Web Services, Inc.   
99     AWS Marketplace  Amazon Web Services, Inc.   

               line_item_legal_entity bill_payer_account_id  \
0   Cloud Infrastructure Services Ltd          981064078967   
1   Cloud Infrastructure Services Ltd          981064078967   
2   Cloud Infrastructure Services Ltd          981064078967   
3   Cloud Infrastructure Services Ltd          981064078967   
4   Cloud Infrastructure Services Ltd          981064078967   
..                                ...                   ...   
95  Cloud Infrastructure Services Ltd          981064078967   
96  Cloud Infrastructure Services Ltd          981064078967   
97  Cloud Infrastructure Services Ltd          981064078967   
98  Cloud Infrastructure Services Ltd          981064078967   
99  Cloud Infrastructure Services Ltd          981064078967   

   line_item_usage_account_id line_item_line_item_type  \
0                273762107511                    Usage   
1                273762107511                    Usage   
2                273762107511                    Usage   
3                273762107511                    Usage   
4                273762107511                    Usage   
..                        ...                      ...   
95               273762107511                    Usage   
96               273762107511                    Usage   
97               273762107511                    Usage   
98               273762107511                    Usage   
99               273762107511                    Usage   

   bill_billing_period_start_date bill_billing_period_end_date  \
0                      2023-05-01                   2023-06-01   
1                      2023-05-01                   2023-06-01   
2                      2023-05-01                   2023-06-01   
3                      2023-05-01                   2023-06-01   
4                      2023-05-01                   2023-06-01   
..                            ...                          ...   
95                     2023-05-01                   2023-06-01   
96                     2023-05-01                   2023-06-01   
97                     2023-05-01                   2023-06-01   
98                     2023-05-01                   2023-06-01   
99                     2023-05-01                   2023-06-01   

   line_item_usage_start_date line_item_usage_end_date  \
0         2023-05-01 00:00:00      2023-05-01 01:00:00   
1         2023-05-01 01:00:00      2023-05-01 02:00:00   
2         2023-05-01 02:00:00      2023-05-01 03:00:00   
3         2023-05-01 03:00:

In [6]:
infra2_cur1['line_item_usage_account_id'].dtype

dtype('O')

In [7]:
infra2_cur1['product_product_name'].unique()

array(['WordPress on Server 2019 + PhpMyAdmin + MySQL',
       'OpenVPN Access Server (5 Connected Devices)',
       'Oracle Linux 8 with Support by Supported Images',
       'ZoneMinder:  Build Your Own Security Camera Solution',
       'Oracle Linux 7 with Support by Supported Images',
       'Reverse Proxy Server using NGINX on Ubuntu',
       'Veeam Backup for AWS Paid Edition'], dtype=object)

## Replicating Infra 2 - Marketplace - Invoice ID 1349042489

### Detail for Consolidated Bill

In [18]:
infra2_cur1.groupby(by='product_product_name')['line_item_unblended_cost'].sum().reset_index()

product_product_name  line_item_unblended_cost
0        OpenVPN Access Server (5 Connected Devices)                   142.848
1    Oracle Linux 7 with Support by Supported Images                   147.100
2    Oracle Linux 8 with Support by Supported Images                   260.900
3         Reverse Proxy Server using NGINX on Ubuntu                     0.144
4                  Veeam Backup for AWS Paid Edition                    24.310
5      WordPress on Server 2019 + PhpMyAdmin + MySQL                    22.260
6  ZoneMinder:  Build Your Own Security Camera So...                     4.470

In [19]:
infra2_cur1['line_item_unblended_cost'].sum()

602.0319999999999

#### <font color='green'>Matches the PDF invoice</font>

### Linked Account Allocation
#### Activity By Account

In [20]:
infra2_cur1.groupby(by='line_item_usage_account_id')['line_item_unblended_cost'].sum()

line_item_usage_account_id
065716851408     73.900
167002523613    358.410
273762107511     22.260
465517541486     71.424
563324065975     71.424
901519325564      4.614
Name: line_item_unblended_cost, dtype: float64

In [21]:
len('065716851408'), len('167002523613'), len('273762107511'), len('465517541486'), len('563324065975'), len('901519325564')

(12, 12, 12, 12, 12, 12)

#### <font color='green'>Matches the PDF invoice</font>

In [22]:
infra2_cur1.groupby(by=['line_item_usage_account_id', 'product_product_name'])['line_item_unblended_cost'].sum()

line_item_usage_account_id  product_product_name                                
065716851408                Oracle Linux 8 with Support by Supported Images          73.900
167002523613                Oracle Linux 7 with Support by Supported Images         147.100
                            Oracle Linux 8 with Support by Supported Images         187.000
                            Veeam Backup for AWS Paid Edition                        24.310
273762107511                WordPress on Server 2019 + PhpMyAdmin + MySQL            22.260
465517541486                OpenVPN Access Server (5 Connected Devices)              71.424
563324065975                OpenVPN Access Server (5 Connected Devices)              71.424
901519325564                Reverse Proxy Server using NGINX on Ubuntu                0.144
                            ZoneMinder:  Build Your Own Security Camera Solution      4.470
Name: line_item_unblended_cost, dtype: float64

## Loading CUR2 from 01_raw

In [23]:
infra2_cur2 = infra2_cur[1][1]()
infra2_cur2.head(100)

identity_line_item_id  bill_invoice_id  \
0  t5fvp5rwqyrvscf3kc2ylwmobdyaazby3yjrbuz4gslxvl...       1366015353   
1  nmuam35xnovmsc7ybj44am7r575ul3obphzy6xln66jykt...       1366015353   
2  kieg2eczb3cyj665s4hklsiw64xx6ldkzl34jgamkyuodv...       1366015353   
3  ikif2qfz7mzuwkwgilejfosjahuhoh5tjhpnrg2uecgan5...       1366015353   

  bill_billing_entity             bill_invoicing_entity  \
0                 AWS  Amazon AWS Serviços Brasil Ltda.   
1                 AWS  Amazon AWS Serviços Brasil Ltda.   
2                 AWS  Amazon AWS Serviços Brasil Ltda.   
3                 AWS  Amazon AWS Serviços Brasil Ltda.   

             line_item_legal_entity bill_payer_account_id  \
0  Amazon AWS Serviços Brasil Ltda.          981064078967   
1  Amazon AWS Serviços Brasil Ltda.          981064078967   
2  Amazon AWS Serviços Brasil Ltda.          981064078967   
3  Amazon AWS Serviços Brasil Ltda.          981064078967   

  line_item_usage_account_id line_item_line_item_type  \
0               072038246234                   Refund   
1               072038246234                      Tax   
2               562293755332                   Refund   
3               562293755332                      Tax   

  bill_billing_period_start_date bill_billing_period_end_date  \
0                     2023-05-01                   2023-06-01   
1                     2023-05-01                   2023-06-01   
2                     2023-05-01                   2023-06-01   
3                     2023-05-01                   2023-06-01   

  line_item_usage_start_date line_item_usage_end_date  line_item_usage_amount  \
0                 2023-05-01               2023-06-01                     0.0   
1                 2023-05-01               2023-06-01                     1.0   
2                 2023-05-01               2023-06-01                     0.0   
3                 2023-05-01               2023-06-01                     1.0   

  line_item_product_code    product_product_name line_item_usage_type  \
0     AWSSupportBusiness  AWS Support (Business)               Dollar   
1     AWSSupportBusiness  AWS Support (Business)               Dollar   
2     AWSSupportBusiness  AWS Support (Business)                  NaN   
3     AWSSupportBusiness  AWS Support (Business)                  NaN   

   line_item_operation  product_from_location_type  product_from_region_code  \
0                  NaN                         NaN                       NaN   
1                  NaN                         NaN                       NaN   
2                  NaN                         NaN                       NaN   
3                  NaN                         NaN                       NaN   

  line_item_tax_type line_item_currency_code  line_item_unblended_cost  
0                NaN                     USD                    -57.61  
1                Tax                     USD                     -7.97  
2                NaN                     USD                     -7.00  
3                Tax                     USD                     -0.96

In [24]:
infra2_cur2['line_item_unblended_cost'].sum()

-73.53999999999999

## Loading CUR3 from 01_raw

In [27]:
infra2_cur3 = infra2_cur[2][1]()
print(infra2_cur3.shape)
infra2_cur3.head(100)

[07/06/23 19:13:12] WARNING  /home/ubuntu/datarain-aws-cur-finops/venv_kedro/lib/python3.10/site-pa ]8;id=1108;file:///usr/lib/python3.10/warnings.py\warnings.py]8;;\:]8;id=942890;file:///usr/lib/python3.10/warnings.py#109\109]8;;\
                             ckages/kedro_datasets/pandas/csv_dataset.py:162: DtypeWarning: Columns                
                             (19) have mixed types. Specify dtype option on import or set                          
                             low_memory=False.                                                                     
                               return pd.read_csv(load_path, **self._load_args)                                    
                                                                                                                   

(764846, 22)


identity_line_item_id  bill_invoice_id  \
0   qmgjvd6gxycmgwrk62rqqkjr63ex5ghqoafbqzurlrpc24...       1343605689   
1   tkd3ppomattn3wyjiak4bp6zlmlo3uaffrnklmk4b3fwdi...       1343605689   
2   fd6dcj5scxyopuenl3n2ighvic2nhuaz3i4fhpanbr3bqf...       1343605689   
3   xldj5f3wd2b4hzzok7zujmonpgcra6uumx4vlvdr6oig7a...       1343605689   
4   t3msytn2lq6tilz4y3pxmqqqqyjhhifkk6qcpkinq2ludy...       1343605689   
..                                                ...              ...   
95  haixin2uwlwnlm5v4ucldjjqrrspevkdnshn4gwewfdffc...       1343605689   
96  pe5a67pil3j22kxjfmztbknf6yquq2kijhvg5bmsrxplba...       1343605689   
97  yg3gumehs3htpiu33fzt5es2wvin2hxjoz2u35kps2tduh...       1343605689   
98  kc2tfbtef7gwxc6sf4f3vl2xq3sdu7gvhzenapvs5edadq...       1343605689   
99  hfyntlujo4vaxvmdnph45dmrxecbpopmkt4ysjqipyh6it...       1343605689   

   bill_billing_entity             bill_invoicing_entity  \
0                  AWS  Amazon AWS Serviços Brasil Ltda.   
1                  AWS  Amazon AWS Serviços Brasil Ltda.   
2                  AWS  Amazon AWS Serviços Brasil Ltda.   
3                  AWS  Amazon AWS Serviços Brasil Ltda.   
4                  AWS  Amazon AWS Serviços Brasil Ltda.   
..                 ...                               ...   
95                 AWS  Amazon AWS Serviços Brasil Ltda.   
96                 AWS  Amazon AWS Serviços Brasil Ltda.   
97                 AWS  Amazon AWS Serviços Brasil Ltda.   
98                 AWS  Amazon AWS Serviços Brasil Ltda.   
99                 AWS  Amazon AWS Serviços Brasil Ltda.   

              line_item_legal_entity bill_payer_account_id  \
0   Amazon AWS Serviços Brasil Ltda.          981064078967   
1   Amazon AWS Serviços Brasil Ltda.          981064078967   
2   Amazon AWS Serviços Brasil Ltda.          981064078967   
3   Amazon AWS Serviços Brasil Ltda.          981064078967   
4   Amazon AWS Serviços Brasil Ltda.          981064078967   
..                               ...                   ...   
95  Amazon AWS Serviços Brasil Ltda.          981064078967   
96  Amazon AWS Serviços Brasil Ltda.          981064078967   
97  Amazon AWS Serviços Brasil Ltda.          981064078967   
98  Amazon AWS Serviços Brasil Ltda.          981064078967   
99  Amazon AWS Serviços Brasil Ltda.          981064078967   

   line_item_usage_account_id line_item_line_item_type  \
0                072038246234                    Usage   
1                368505959464                    Usage   
2                543280990914                    Usage   
3                368505959464                    Usage   
4                420065387364                    Usage   
..                        ...                      ...   
95               901519325564                    Usage   
96               684481376085                    Usage   
97               901519325564                    Usage   
98               494939165758                    Usage   
99               072038246234                    Usage   

   bill_billing_period_start_date bill_billing_period_end_date  \
0                      2023-05-01                   2023-06-01   
1                      2023-05-01                   2023-06-01   
2                      2023-05-01                   2023-06-01   
3                      2023-05-01                   2023-06-01   
4                      2023-05-01                   2023-06-01   
..                            ...                          ...   
95                     2023-05-01                   2023-06-01   
96                     2023-05-01                   2023-06-01   
97                     2023-05-01                   2023-06-01   
98                     2023-05-01                   2023-06-01   
99                     2023-05-01                   2023-06-01   

   line_item_usage_start_date line_item_usage_end_date  \
0         2023-05-15 16:00:00      2023-05-15 17:00:00   
1         2023-05-25 20:00:00      2023-05-25 21:00:00   
2         202

In [28]:
infra2_cur3.groupby(by=['line_item_usage_account_id', 'product_product_name', 'line_item_currency_code'])['line_item_unblended_cost'].sum()

line_item_usage_account_id  product_product_name                line_item_currency_code
065716851408                AWS Config                          USD                          0.282000
                            AWS Key Management Service          USD                          0.000150
                            Amazon Elastic Compute Cloud        USD                        197.672197
                            Amazon Simple Notification Service  USD                          0.000001
                            Amazon Simple Queue Service         USD                          0.000004
                                                                                              ...    
988350742118                Amazon Elastic Compute Cloud        USD                         11.680824
                            Amazon Relational Database Service  USD                         18.545119
                            Amazon Simple Storage Service       USD                          7.2

In [29]:
a = pd.DataFrame({'a': [1, 2, 3], 'b': [1, 2, 3]})
b = pd.DataFrame({'a': [4, 5, 6], 'b': [4, 5, 6]})

pd.concat([a,b])

a  b
0  1  1
1  2  2
2  3  3
0  4  4
1  5  5
2  6  6